compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -541,16 +553,84 @@ private BType getFiniteTypeMatchWithIntLiteral(BLangLiteral literalExpr, BFinite
             literalExpr.value = String.valueOf(literalValue);
             return type;
         }
-        return symTable.intType;
+
+        if (literalValue instanceof Double) {
+            return symTable.floatType;
+        } else if (literalValue instanceof String) {
+            return symTable.decimalType;
+        } else {
+            return symTable.intType;
+        }
+    }
+
+    private BType clonedIntTypeChecker(BLangLiteral literalExpr, Object literalValue, BType expType,
+                                       AnalyzerData data) {
+        boolean prevNonErrorLoggingCheck = data.nonErrorLoggingCheck;
+        data.nonErrorLoggingCheck = true;
+        int prevErrorCount = this.dlog.errorCount();
+        this.dlog.resetErrorCount();
+        this.dlog.mute();
+
+        BType exprCompatibleType = getIntegerLiteralType(nodeCloner.cloneNode(literalExpr), literalValue, expType,
+                data);
+        data.nonErrorLoggingCheck = prevNonErrorLoggingCheck;
+        this.dlog.setErrorCount(prevErrorCount);
+
+        if (!prevNonErrorLoggingCheck) {
+            this.dlog.unmute();
+        }
+        return exprCompatibleType;
+    }
+
+    private BType clonedCheckForCompatibleLiteralType(BFiniteType finiteType, BLangLiteral literalExpr,
+                                                      Object literalValue, AnalyzerData data) {
+        BType resIntType = symTable.semanticError;
+        for (BLangExpression valueExpr : finiteType.getValueSpace()) {
+            BType tempResType = clonedIntTypeChecker(literalExpr, literalValue, valueExpr.getBType(), data);
+            if (tempResType != symTable.semanticError) {
+                resIntType = tempResType;
+                break;
+            }
+        }
+        return resIntType;
+    }
+
+    private BType checkIfLiteralIsOutOfRange(BFiniteType finiteType, BLangLiteral literalExpr, Object literalValue,
+                                             AnalyzerData data) {
+        BType originalLiteralType = literalExpr.getBType();
+        BType compatibleType = clonedCheckForCompatibleLiteralType(finiteType, literalExpr, literalValue, data);
+        if (compatibleType == symTable.semanticError) {
+            String literalType;
+            if (originalLiteralType.tag == TypeTags.INT) {
+                literalType = "Integer";
+            } else {
+                literalType = "Float";
+            }
```

Comment:
See comments re: sticking to the actual type name.

Reply:
This function only gets called if the expected type is finite. So we won't have sub types there right.